File: oop/account-exception-app/src/com/techlabs/account/Account.java
package com.techlabs.account;

public class Account {
	private int accNo;
	private String name;
	private double balance;
	
	private static final double MIN_BALANCE = 500.0;
	
	public Account(int accNo, String name, double balance)
	{
		this.accNo = accNo;
		this.name = name;
		this.balance = balance;
	}
	
	public int getAccNo()
	{
		return this.accNo;
	}
	
	public String getName()
	{
		return this.name;
	}
	
	public double getBalance()
	{
		return this.balance;
	}
	
	public void deposite(double amount)
	{
		if (amount < 0.0)
			return;
		this.balance += amount;
	}
	
	public void withdraw(double amount)
	{
		if (amount < 0.0)
			return;
		if (this.balance - amount < MIN_BALANCE)
			throw new RuntimeException("No funds to withdraw");
		this.balance -= amount;
	}
}

File: go/src/account-app/account/account.go
package account

// Account account structure
type Account struct {
	accNo   int64
	name    string
	balance float64
}

func NewAccount(accNo int64, name string, balance float64) *Account {
	if balance < 0 {
		panic("Balance should be positive")
	}
	return &Account{
		accNo:   accNo,
		name:    name,
		balance: balance,
	}
}

func (acc *Account) Balance() float64 {
	return acc.balance
}

func (acc *Account) Name() string {
	return acc.name
}

func (acc *Account) AccNo() int64 {
	return acc.accNo
}

func (acc *Account) Deposit(amount float64) {
	acc.balance += amount
}

func (acc *Account) Withdraw(amount float64) {
	if amount < 0 || acc.balance-amount < 500 {
		panic("Given amount cant be withdrawn")
	}
	acc.balance -= amount
}
